File: applications/luci-app-statistics/luasrc/model/cbi/luci_statistics/network.lua
-- Copyright 2008 Freifunk Leipzig / Jo-Philipp Wich <jow@openwrt.org>
-- Licensed to the public under the Apache License 2.0.

require "luci.util"
local fs = require "nixio.fs"

m = Map("luci_statistics",
	translate("Network Plugin Configuration"),
	translate(
		"The network plugin provides network based communication between " ..
		"different collectd instances. Collectd can operate both in client " ..
		"and server mode. In client mode locally collected data is " ..
		"transferred to a collectd server instance, in server mode the " ..
		"local instance receives data from other hosts."
	))

-- collectd_network config section
s = m:section( NamedSection, "collectd_network", "luci_statistics" )

-- collectd_network.enable
enable = s:option( Flag, "enable", translate("Enable this plugin") )
enable.default = 0

local have_encrypted_network_support = luci.util.checklib("/usr/lib/collectd/network.so", "libgcrypt.so")
local f

-- collectd_network_listen config section (Listen)
listen = m:section( TypedSection, "collectd_network_listen",
	translate("Listener interfaces"),
	translate(
		"This section defines on which interfaces collectd will wait " ..
		"for incoming connections."
	))
listen.addremove = true
listen.anonymous = true

-- collectd_network_listen.host
listen_host = listen:option( Value, "host", translate("Listen host") )
listen_host.default = "0.0.0.0"

-- collectd_network_listen.port
listen_port = listen:option( Value, "port", translate("Listen port") )
listen_port.default   = 25826
listen_port.isinteger = true
listen_port.optional  = true

if have_encrypted_network_support then
	securitylevel = listen:option(ListValue, "security_level", translate("Security Level"))
	securitylevel:value("None", translate("No encryption"))
	securitylevel:value("Encrypt", translate("Encrypted data only"))
	securitylevel:value("Sign", translate("Signed and encrypted data only"))
	securitylevel.default = "None"
	authfile = listen:option(Value, "auth_file", translate("Auth file"), translate("File containing username:password pairs to allow"))
	authfile:depends("security_level", "Encrypt")
	authfile:depends("security_level", "Sign")
	authfile.placeholder = "/etc/collectd/authfile"

	t = listen:option(TextValue, "auths", translate("Auth File"), translate("username:password pairs allowed by this listener"))
	t:depends("security_level", "Encrypt")
	t:depends("security_level", "Sign")
	t.rmempty = true
	t.rows = 10
	function t.cfgvalue(self, section)
		local afile = authfile:cfgvalue(section) or "/etc/collectd/authfile"
		return fs.readfile(afile) or ""
	end

	function t.write(self, section, value)
		local afile = authfile:cfgvalue(section) or "/etc/collectd/authfile"
		if value then
                        fs.writefile(afile, value:gsub("\r\n", "\n"))
		else
			fs.writefile(afile, "")
		end
	end

	function t.remove(self, section)
		local afile = authfile:cfgvalue(section) or "/etc/collectd/authfile"
		fs.writefile(afile, "")
	end
end

-- collectd_network_server config section (Server)
server = m:section( TypedSection, "collectd_network_server",
	translate("server interfaces"),
	translate(
		"This section defines to which servers the locally collected " ..
		"data is sent to."
	))
server.addremove = true
server.anonymous = true

-- collectd_network_server.host
server_host = server:option( Value, "host", translate("Server host") )
server_host.default = "0.0.0.0"

-- collectd_network_server.port
server_port = server:option( Value, "port", translate("Server port") )
server_port.default   = 25826
server_port.isinteger = true
server_port.optional  = true

if have_encrypted_network_support then
	securitylevel = server:option(ListValue, "security_level", translate("Security Level"))
	securitylevel:value("None", translate("No encryption"))
	securitylevel:value("Encrypt", translate("Encrypted data only"))
	securitylevel:value("Sign", translate("Signed and encrypted data only"))
	username = server:option(Value, "username", translate("Username"))
	username:depends("security_level", "Encrypt")
	username:depends("security_level", "Sign")
	password = server:option(Value, "password", translate("Password"))
	password:depends("security_level", "Encrypt")
	password:depends("security_level", "Sign")
	password.password = true
	resolveinterval = server:option(Value, "resolve_interval", translate("Resolve Interval"), translate("Interval (in seconds) in which to redo resolution of DNS names."))
	resolveinterval.datatype = "uinteger"
	resolveinterval.optional = true
end

-- collectd_network.timetolive (TimeToLive)
ttl = s:option( Value, "TimeToLive", translate("TTL for network packets") )
ttl.default   = 128
ttl.isinteger = true
ttl.optional  = true
ttl:depends( "enable", 1 )

-- collectd_network.forward (Forward)
forward = s:option( Flag, "Forward", translate("Forwarding between listen and server addresses") )
forward.default  = 0
forward.optional = true
forward:depends( "enable", 1 )

-- collectd_network.cacheflush (CacheFlush)
cacheflush = s:option( Value, "CacheFlush",
	translate("Cache flush interval"), translate("Seconds") )
cacheflush.default   = 86400
cacheflush.isinteger = true
cacheflush.optional  = true
cacheflush:depends( "enable", 1 )


return m, f

File: applications/luci-app-statistics/htdocs/luci-static/resources/view/statistics/plugins/network.js
'use strict';
'require baseclass';
'require form';

return baseclass.extend({
	title: _('Network Plugin Configuration'),
	description: _('The network plugin provides network based communication between different collectd instances. Collectd can operate both in client and server mode. In client mode locally collected data is transferred to a collectd server instance, in server mode the local instance receives data from other hosts.'),

	addFormOptions: function(s) {
		var o, ss;

		o = s.option(form.Flag, 'enable', _('Enable this plugin'));
		o.default = '0';

		o = s.option(form.Value, 'TimeToLive', _('TTL for network packets'));
		o.placeholder = '1';
		o.datatype = 'range(1, 255)';
		o.depends('enable', '1');

		o = s.option(form.Value, 'MaxPacketSize', _('Maximum packet size'), _('Set the maximum size for datagrams sent over the network'));
		o.placeholder = '1452';
		o.datatype = 'range(1024, 65535)';
		o.depends('enable', '1');

		o = s.option(form.Flag, 'Forward', _('Enable forwarding'), _('Forwarding between listen and server addresses'));
		o.depends('enable', '1');

		o = s.option(form.Flag, 'ReportStats', _('Enable statistics'), _('Create statistics about the network plugin itself'));
		o.depends('enable', '1');

		o = s.option(form.SectionValue, '__listeners', form.TableSection, 'collectd_network_listen');
		o.title = _('Listener interfaces');
		o.description = _('This section defines on which interfaces collectd will wait for incoming connections.');
		o.depends('enable', '1');

		ss = o.subsection;
		ss.anonymous = true;
		ss.addremove = true;

		o = ss.option(form.Value, 'host', _('Listen host'));
		o.default = '0.0.0.0';
		o.datatype = 'ipaddr("nomask")';

		o = ss.option(form.Value, 'port', _('Listen port'));
		o.default = '25826';
		o.datatype = 'port';
		//o.optional = true;

		o = s.option(form.SectionValue, '__servers', form.TableSection, 'collectd_network_server');
		o.title = _('Server interfaces');
		o.description = _('This section defines to which servers the locally collected data is sent to.');
		o.depends('enable', '1');

		ss = o.subsection;
		ss.anonymous = true;
		ss.addremove = true;

		o = ss.option(form.Value, 'host', _('Server host'));
		o.default = '0.0.0.0';
		o.datatype = 'or(hostname,ipaddr("nomask"))';

		o = ss.option(form.Value, 'port', _('Server port'));
		o.default = '25826';
		o.datatype = 'port';
		//o.optional = true;
	},

	configSummary: function(section) {
		return _('Network communication enabled');
	}
});
